### PR TITLE
GH-43299: [Release][Packaging] Only include pyarrow folder  when finding packages on setuptools

### PR DESCRIPTION
**Pull request by raulcd**
### Rationale for this change

Currently we include everything when building wheels, see:
```
$ pip install pyarrow
Collecting pyarrow
  Downloading pyarrow-17.0.0-cp310-cp310-manylinux_2_28_x86_64.whl (39.9 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 39.9/39.9 MB 33.8 MB/s eta 0:00:00
Collecting numpy>=1.16.6
  Using cached numpy-2.0.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (19.3 MB)
Installing collected packages: numpy, pyarrow
Successfully installed numpy-2.0.0 pyarrow-17.0.0
(test-env)  $ ls test-env/lib/python3.10/site-packages/
benchmarks/                  distutils-precedence.pth     numpy-2.0.0.dist-info/       pip-22.0.2.dist-info/        pyarrow-17.0.0.dist-info/    setuptools-59.6.0.dist-info/ 
cmake_modules/               examples/                    numpy.libs/                  pkg_resources/               scripts/                     
_distutils_hack/             numpy/                       pip/                         pyarrow/                     setuptools/    
```

### What changes are included in this PR?

Use `include` as seen here: https://setuptools.pypa.io/en/latest/userguide/package_discovery.html#finding-simple-packages

### Are these changes tested?

Will check via the build wheel on CI

### Are there any user-facing changes?

No and yes :)
We will remove unnecessary files
* GitHub Issue: #43299

### CODE REVIEW

**Comment by raulcd** [comment]
@github-actions crossbow submit -g python wheel

**Comment by raulcd** [comment]
There seemed to be a minor issue:
```
$ ls -lrt
total 77296
drwxr-xr-x 2 raulcd raulcd     4096 jul 18 15:11 pyarrow-18.0.0.dev15.dist-info
drwxr-xr-x 2 raulcd raulcd     4096 jul 18 15:11 pyarrow.
drwxr-xr-x 9 raulcd raulcd     4096 jul 18 15:11 pyarrow
$ ls -lrt pyarrow./
total 0
```

**Comment by raulcd** [comment]
@github-actions crossbow submit wheel-manylinux-2-28-cp310-amd64

**Comment by raulcd** [comment]
@github-actions crossbow submit wheel-manylinux-2-28-cp310-amd64

**Comment by raulcd** [comment]
@github-actions crossbow submit wheel-manylinux-2-28-cp310-amd64

**Comment by raulcd** [comment]
@github-actions crossbow submit wheel-manylinux-2-28-cp310-amd64

**Comment by raulcd** [comment]
@github-actions crossbow submit wheel-manylinux-2-28-cp310-amd64

File: python/pyproject.toml [hunk 1]
```diff
@@ -73,7 +73,9 @@ zip-safe=false
 include-package-data=true
 
 [tool.setuptools.packages.find]
-where = ["."]
+include = ["pyarrow"]
+exclude = ["pyarrow/tests", "pyarrow."]
```
Comment (raulcd) [inline python/pyproject.toml:77]: this doesn't seem to have any effect, @jorisvandenbossche @pitrou any idea?
The other stray directories are filtered from the wheel now

**Comment by timkpaine** [comment]
was something wrong with https://github.com/apache/arrow/pull/43281?

**Comment by raulcd** [comment]
> was something wrong with #43281 from 3 days ago?

Sorry, I didn't see that issue, it would be good to mark the second one as a duplicate. I also think that as the issue was introduced with a wrongly configured pyproject.toml it might be worth also fixing it there

File: python/pyproject.toml (hunk 1 above)
Comment (anjakefala) [reply python/pyproject.toml:77]: I believe the syntax is `pyarrow.tests`. What does `pyarrow.` refer to?

Comment (raulcd) [reply python/pyproject.toml:77]: there is a folder included in the wheel `pyarrow.` which is empty and I was trying to remove it. I'll try with `pyarrow.tests`

**Comment by raulcd** [comment]
@github-actions crossbow submit wheel-manylinux-2-28-cp310-amd64

File: python/pyproject.toml (hunk 1 above)
Comment (raulcd) [reply python/pyproject.toml]: it doesn't remove tests either

Comment (pitrou) [reply python/pyproject.toml]: We don't care about removing the tests, they are comparatively small and can serve to test PyArrow on a third-party machine.
```console
$ du -hs venv-3.10/lib/python3.10/site-packages/pyarrow
187M	venv-3.10/lib/python3.10/site-packages/pyarrow
$ du -hs venv-3.10/lib/python3.10/site-packages/pyarrow/tests/
4,0M	venv-3.10/lib/python3.10/site-packages/pyarrow/tests/
```

Comment (raulcd) [reply python/pyproject.toml]: what about cpp and header files, should we do something like pandas here:
https://github.com/pandas-dev/pandas/blob/main/pyproject.toml#L131

Comment (raulcd) [reply python/pyproject.toml]: Currently the files under `pyarrow/src/arrow/python`  are included on the wheel

Comment (pitrou) [reply python/pyproject.toml]: I'll stress that the problem we're trying to solve is that installing a PyArrow wheel creates top-level directories _outside_ of the `pyarrow` source tree. This is the urgency. Cleaning up the contents of `pyarrow` is a separate task, much less critical.

Comment (raulcd) [reply python/pyproject.toml]: I guess those are negligible and we still require to distribute them on the source distitrbution:
```
pyarrow/src $ du -hs .
728K	.
```

Comment (raulcd) [reply python/pyproject.toml]: > I'll stress that the problem we're trying to solve is that installing a PyArrow wheel creates top-level directories _outside_ of the `pyarrow` source tree.

ok, then, this is solved by this PR, are you ok with the current changes @pitrou ?

**Review by pitrou — APPROVED** [review]
LGTM, but you might want to wait for @jorisvandenbossche 's comeback tomorrow

**Review by anjakefala — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
@raulcd thanks for working on this! We should have tested that on the original PR .. 

There is still some code in `setup.py` to define `packages`, which can be cleaned up. Given that this code is still there, and from looking back at the discussion in https://github.com/apache/arrow/pull/41041, I think we assumed that this would still work (i.e. moved most info to pyproject.toml, but leave `packages` definition in setup.py). 
Because for example we have the `PYARROW_INSTALL_TESTS` env variable to allow removing the tests from the packages (but it seems we only use that in our CI turned on, and not turned off .. so this "feature" also broke in 17.0 I assume)

**Comment by jorisvandenbossche** [comment]
> I am unsure on why an empty directory with `pyarrow.` is being added

That's a longer standing issue that I am also not entirely sure about (that goes back to 6.0.0 and is linux only AFAIK). I thought that I opened an issue about that at some point, but it seems not. Could also be related to something in a dependency like setuptools (my first guess is that it might have something to do with creating a directory for libs that goes wrong (and is not needed in our case as we put all .so files in the package itself)

**Comment by raulcd** [comment]
> There is still some code in `setup.py` to define `packages`, which can be cleaned up.

sure, @jorisvandenbossche do you prefer to do this cleanup on this PR?

**Comment by jorisvandenbossche** [comment]
While it's not critical for this PR to fix the issue (since it's just deal old code now), I think I would prefer to just include it, to not further scatter changes related to this.

Also the `PYARROW_INSTALL_TESTS` env variable is documented, so that would be good to remove from the docs in a PR that gets backported in case there is a 17.0.1 release.

**Comment by jorisvandenbossche** [comment]
This is also assuming here that we are fine with actually removing the `PYARROW_INSTALL_TESTS=0` option.

I don't think many people are using this (I verified that conda-forge is not using it, they have there own handling of providing a separate `pyarrow-tests` output). I found only a few occurrences through a github code search (https://github.com/altlinux/specs/blob/f271ff0e6e3abfd44ce06c1b88655570b58060ec/a/arrow/arrow.spec#L532
https://github.com/ehenry2/minimal-pyarrow/blob/a81244cce71b84d131a547cc618ba065eb1b658f/build.sh#L55
https://github.com/mamckee/azurelinux/blob/253b6a9ff1f1708c2a55f7fde8292340374f425f/SPECS/libarrow/libarrow.spec#L351).

I assume that in principle we could still have some logic in setup.py to override the included files from pyproject.toml, but that's probably not worth it?

**Comment by raulcd** [comment]
@github-actions crossbow submit wheel-manylinux-2-28-cp310-amd64

**Comment by jorisvandenbossche** [comment]
One other question: also the sdist started to include more content. It's not entirely clear to me if that is also controlled by the change you made in pyproject.toml, or if that still needs a change to MANIFEST.in as well (although that file was not changed)

(do we have a nightly build to trigger to test this?, `python-sdist` doesn't seem to include the resulting sdist as an artifact on the workflow)

**Comment by jorisvandenbossche** [comment]
> Will check via the build wheel on CI

In _theory_ we could also add a test for this (unpacking the resulting archive, ensuring that only the expected directories are present), although that might be more work than worth it

**Comment by pitrou** [comment]
> One other question: also the sdist started to include more content.

It's not a problem for a sdist to contain the entire source tree, including non-package contents. The goal is to have an artifact that can be used to build PyArrow (which is why it has to contain `CMakeLists.txt`, `setup.py`, etc.)

> In theory we could also add a test for this (unpacking the resulting archive, ensuring that only the expected directories are present), although that might be more work than worth it

We could probably do this indeed. It shouldn't be a lot of work.

**Comment by pitrou** [comment]
Quick and dirty POC:
```pycon
>>> f = zipfile.ZipFile("pyarrow-18.0.0.dev115-cp312-cp312-manylinux_2_28_aarch64.whl")
>>> outliers = [info.filename for info in f.filelist if not re.match(r'(pyarrow/|pyarrow-[-.\w\d]+\.dist-info/)', info.filename)]
>>> assert not outliers, f"Unexpected contents in wheel: {sorted(outliers)}"
Traceback (most recent call last):
  Cell In[24], line 1
    assert not outliers, f"Unexpected contents in wheel: {sorted(outliers)}"
AssertionError: Unexpected contents in wheel: ['benchmarks/', 'benchmarks/__init__.py', 'benchmarks/array_ops.py', 'benchmarks/common.py', 'benchmarks/convert_builtins.py', 'benchmarks/convert_pandas.py', 'benchmarks/io.py', 'benchmarks/microbenchmarks.py', 'benchmarks/parquet.py', 'benchmarks/streaming.py', 'cmake_modules/', 'cmake_modules/AWSSDKVariables.cmake', 'cmake_modules/BuildUtils.cmake', 'cmake_modules/DefineOptions.cmake', 'cmake_modules/FindAWSSDKAlt.cmake', 'cmake_modules/FindAzure.cmake', 'cmake_modules/FindBrotliAlt.cmake', 'cmake_modules/FindClangTools.cmake', 'cmake_modules/FindGTestAlt.cmake', 'cmake_modules/FindInferTools.cmake', 'cmake_modules/FindLLVMAlt.cmake', 'cmake_modules/FindOpenSSLAlt.cmake', 'cmake_modules/FindProtobufAlt.cmake', 'cmake_modules/FindPython3Alt.cmake', 'cmake_modules/FindRapidJSONAlt.cmake', 'cmake_modules/FindSQLite3Alt.cmake', 'cmake_modules/FindSnappyAlt.cmake', 'cmake_modules/FindThriftAlt.cmake', 'cmake_modules/Findc-aresAlt.cmake', 'cmake_modules/FindgRPCAlt.cmake', 'cmake_modules/FindgflagsAlt.cmake', 'cmake_modules/FindglogAlt.cmake', 'cmake_modules/FindjemallocAlt.cmake', 'cmake_modules/Findlibrados.cmake', 'cmake_modules/Findlz4Alt.cmake', 'cmake_modules/FindorcAlt.cmake', 'cmake_modules/Findre2Alt.cmake', 'cmake_modules/Findutf8proc.cmake', 'cmake_modules/FindzstdAlt.cmake', 'cmake_modules/GandivaAddBitcode.cmake', 'cmake_modules/SetupCxxFlags.cmake', 'cmake_modules/ThirdpartyToolchain.cmake', 'cmake_modules/UseCython.cmake', 'cmake_modules/Usevcpkg.cmake', 'cmake_modules/aws_sdk_cpp_generate_variables.sh', 'cmake_modules/san-config.cmake', 'cmake_modules/snappy.diff', 'examples/', 'examples/dataset/', 'examples/dataset/write_dataset_encrypted.py', 'examples/flight/', 'examples/flight/client.py', 'examples/flight/middleware.py', 'examples/flight/server.py', 'examples/parquet_encryption/', 'examples/parquet_encryption/sample_vault_kms_client.py', 'pyarrow./', 'scripts/', 'scripts/run_emscripten_tests.py', 'scripts/test_imports.py', 'scripts/test_leak.py']
```

**Comment by jorisvandenbossche** [comment]
@raulcd do you have time to update this?

**Comment by raulcd** [comment]
> @raulcd do you have time to update this?

to add the test to validate the contents? Sure, I can try to find some time to give that a try in the next days

**Comment by raulcd** [comment]
@github-actions crossbow submit wheel-manylinux-2-28-cp310-amd64

**Comment by raulcd** [comment]
@github-actions crossbow submit wheel-manylinux-2-28-cp310-amd64

**Comment by raulcd** [comment]
@github-actions crossbow submit wheel-manylinux-2-28-cp310-amd64

**Comment by raulcd** [comment]
@github-actions crossbow submit wheel-windows-cp39-amd64

**Comment by raulcd** [comment]
@github-actions crossbow submit wheel-windows-cp39-amd64

**Comment by raulcd** [comment]
@github-actions crossbow submit -g wheel

**Comment by raulcd** [comment]
@jorisvandenbossche the check is working on macOS, manylinux and Windows wheels tests now. If you can review again. Thanks!

**Comment by pitrou** [comment]
I think I've found the source of the stray `pyarrow.` directory. It's because of https://github.com/pypa/auditwheel/issues/488 and the `-L .` option.

**Comment by pitrou** [comment]
@github-actions crossbow submit wheel*cp312*

**Comment by pitrou** [comment]
@github-actions crossbow submit -g wheel

**Comment by pitrou** [comment]
CI failures are unrelated. @jorisvandenbossche Do you want to take a last look?

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
> I think I've found the source of the stray `pyarrow.` directory. It's because of [pypa/auditwheel#488](https://github.com/pypa/auditwheel/issues/488) and the `-L .` option.

Nice catch!

**Comment by jorisvandenbossche** [comment]
Thanks @raulcd

**Pull request merged**